### PR TITLE
uv: Fix extras normalization mismatch in pyproject.toml updates

### LINKED ISSUE #14415: dependabot uv - fails when "extras" have spaces - "Dependency file content did not change"
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Package ecosystem

uv

### Package manager version

uv

### Language version

python 3.13

### Manifest location and content before the Dependabot update

[Archive.zip](https://github.com/user-attachments/files/25903953/Archive.zip)

### dependabot.yml content

version: 2
updates:
  - package-ecosystem: uv
    schedule:
      interval: weekly
      day: "thursday"
      time: "14:15"
    commit-message:
      prefix: '(test-repo) '
    
    directory: /
    open-pull-requests-limit: 1
    patterns: ["*"]
    groups:
      all-updates:
        patterns: ["*"]
    insecure-external-code-execution: allow

### Updated dependency

"aws-lambda-powertools[aws-sdk, tracer]~=3.24.0"  to "aws-lambda-powertools[aws-sdk, tracer]~=3.25.0"

### What you expected to see, versus what you actually saw

Expected:

Dependabot should update the package to 3.25.0 in both pyproject.toml and uv.lock

Actual:

"Dependency file content did not change" in logs and no PR raised

### Native package manager behavior

All works fine

### Images of the diff or a link to the PR, issue, or logs

[log.txt](https://github.com/user-attachments/files/25904057/log.txt)

### Smallest manifest that reproduces the issue

[manifest.zip](https://github.com/user-attachments/files/25904099/manifest.zip)

### LINKED ISSUE #14416: dependabot uv - fails when "extras" are not in alphabetical order - "Dependency file content did not change"
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Package ecosystem

uv

### Package manager version

uv

### Language version

python 3.13

### Manifest location and content before the Dependabot update

[Archive.zip](https://github.com/user-attachments/files/25904195/Archive.zip)

### dependabot.yml content

version: 2
updates:
  - package-ecosystem: uv
    schedule:
      interval: weekly
      day: "thursday"
      time: "14:16"
    commit-message:
      prefix: '(test-repo) '
    
    directory: /
    open-pull-requests-limit: 1
    patterns: ["*"]
    groups:
      all-updates:
        patterns: ["*"]
    insecure-external-code-execution: allow

### Updated dependency

"aws-lambda-powertools[tracer,aws-sdk]~=3.24.0" to "aws-lambda-powertools[tracer,aws-sdk]~=3.24.5"

### What you expected to see, versus what you actually saw

Expected:

Dependabot should update pyproject.tonl and uv.lock in a PR

Actual:

Dependabot error: "Dependency file content did not change"

### Native package manager behavior

All OK

### Images of the diff or a link to the PR, issue, or logs

[log.txt](https://github.com/user-attachments/files/25904371/log.txt)

### Smallest manifest that reproduces the issue

[manifest.zip](https://github.com/user-attachments/files/25904456/manifest.zip)

### PR DESCRIPTION
**Pull request by awinogradov**
## What

Fix `DependencyFileContentNotChanged` error when updating dependencies that have extras with spaces or in non-alphabetical order in `pyproject.toml`.

## Why

When a `pyproject.toml` contains a dependency like:
```toml
"aws-lambda-powertools[aws-sdk, tracer]~=3.24.0"
```

The Python parser sorts extras alphabetically and the Ruby normalizer joins them without spaces, producing the canonical name `aws-lambda-powertools[aws-sdk,tracer]`. The `LockFileUpdater#replace_dep` method builds a regex from this normalized name to find and replace the version in the source file — but the regex `"aws-lambda-powertools[aws-sdk,tracer]"` can never match `"aws-lambda-powertools[aws-sdk, tracer]"` (with space) or `"aws-lambda-powertools[tracer, aws-sdk]"` (different order).

The `gsub` silently returns unchanged content, and line 130 raises `DependencyFileContentNotChanged`.

## How

Updated `escape_package_name` in `LockFileUpdater` to detect extras in dependency names and build a flexible regex using lookaheads:

- Each extra becomes a lookahead `(?=[^\]]*extra_name)` — order-independent matching
- The bracket content matches `[^\]]+` — allows any whitespace around commas
- Falls back to existing behavior for names without extras
- Does NOT change `NameNormaliser` (shared across 7+ callers in pip/uv/python ecosystems)

## Tests

3 new RSpec test cases added to `lock_file_updater_spec.rb`:
- Extras with spaces in source: `[aws-sdk, tracer]` ✅
- Extras in non-alphabetical order: `[tracer, aws-sdk]` ✅
- Extras matching exactly (regression): `[aws-sdk,tracer]` ✅

Full test suite: **65 examples, 0 failures**

## Reproduction

`pyproject.toml`:
```toml
[project]
dependencies = [
    "aws-lambda-powertools[aws-sdk, tracer]~=3.24.0",
]
```

Before: Dependabot log shows `Dependency file content did not change`
After: Version correctly updated to `~=3.25.0` preserving original extras formatting

Fixes #14415
Fixes #14416

### CODE REVIEW

**Comment by v-thavaahariharangit** [comment]
Could you please resolve the copilot comments above pls

File: uv/lib/dependabot/uv/file_updater/lock_file_updater.rb [hunk 1]
```diff
@@ -460,8 +460,22 @@ def lock_options_fingerprint(options)
...
+          match = name_str.match(/\A([^\[]+)\[([^\]]+)\]\z/)
+
+          # Handle extras: "pkg[extra1,extra2]" needs flexible matching for
+          # whitespace around commas and any ordering of extras in source file
+          if match
+            base = Regexp.escape(T.must(match[1])).gsub(/\\[-_.]/, "[-_.]")
+            extras = T.must(match[2]).split(",").map(&:strip)
+            extras_patterns = extras.map { |e| Regexp.escape(e).gsub(/\\[-_.]/, "[-_.]") }
+            # Use lookaheads so extras match in any order
+            lookaheads = extras_patterns.map { |e| "(?=[^\\]]*#{e})" }.join
```
Comment (v-thavaahariharangit) [inline uv/lib/dependabot/uv/file_updater/lock_file_updater.rb:473]: 🟡 **Medium: Lookahead can false-positive on partial extra names**

The lookahead `(?=[^\]]*#{e})` will match a substring of a longer extra name. For example, if the dependency has extras `[bar,baz]` and the source file has `[barbaz,qux]`, the regex would incorrectly match `bar` inside `barbaz`.

Consider adding comma/bracket boundaries so extras are matched as whole tokens:

```ruby
lookaheads = extras_patterns.map { |e| "(?=[^\\]]*(?:,\s*|\[)#{e}(?:,|\]))" }.join
```

This ensures each extra is preceded by `[` or `,` and followed by `,` or `]`.

File: uv/spec/dependabot/uv/file_updater/lock_file_updater_spec.rb [hunk 2]
```diff
@@ -1339,6 +1339,129 @@
         end
       end
     end
+
+    context "with extras that have spaces in source file" do
```
Comment (v-thavaahariharangit) [inline uv/spec/dependabot/uv/file_updater/lock_file_updater_spec.rb:1343]: 🟡 **Medium: Test duplication across three contexts**

The `dependency`, `new_req`, and `old_req` `let` blocks are copy-pasted identically in all three new contexts. Consider wrapping them in a shared parent context where only `content` and expectations vary:

```ruby
context "with extras in dependency name" do
  let(:dependency) do
    Dependabot::Dependency.new(
      name: "aws-lambda-powertools[aws-sdk,tracer]",
      # ... shared setup ...
    )
  end
  let(:new_req) { { requirement: "~=3.25.0" } }
  let(:old_req) { { requirement: "~=3.24.0" } }

  context "with spaces in source file" do
    let(:content) { ... }
    it "matches extras despite spaces" do ... end
  end
  # ... other sub-contexts with only content varying
end
```

🟢 **Low: Missing edge-case coverage**
- No test for a **single-extra** case (e.g., `pkg[extra]`) which exercises the bracket matching differently.
- No test for extras containing `.` or `_` separators (e.g., `my_extra` in the dependency vs `my-extra` in source) to verify PEP 503 normalization works within extras too.
